File: src/test/java/uk/q3c/krail/core/navigate/sitemap/DefaultSitemapServiceTest.java
/*
 *
 *  * Copyright (c) 2016. David Sowerby
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 *  * the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 *  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 *  * specific language governing permissions and limitations under the License.
 *
 */
package uk.q3c.krail.core.navigate.sitemap;

import com.google.inject.Inject;
import com.mycila.testing.junit.MycilaJunitRunner;
import com.mycila.testing.plugin.guice.GuiceContext;
import com.vaadin.server.VaadinService;
import org.apache.commons.configuration.ConfigurationException;
import org.apache.commons.configuration.HierarchicalINIConfiguration;
import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import uk.q3c.krail.core.config.ApplicationConfigurationModule;
import uk.q3c.krail.core.config.ConfigKeys;
import uk.q3c.krail.core.eventbus.EventBusModule;
import uk.q3c.krail.core.guice.uiscope.UIScopeModule;
import uk.q3c.krail.core.guice.vsscope.VaadinSessionScopeModule;
import uk.q3c.krail.core.navigate.NavigationModule;
import uk.q3c.krail.core.navigate.sitemap.DefaultSitemapServiceTest.TestDirectSitemapModule;
import uk.q3c.krail.core.navigate.sitemap.set.MasterSitemapQueue;
import uk.q3c.krail.core.services.Service;
import uk.q3c.krail.core.services.ServiceStatus;
import uk.q3c.krail.core.services.ServicesModel;
import uk.q3c.krail.core.services.ServicesModule;
import uk.q3c.krail.core.shiro.DefaultShiroModule;
import uk.q3c.krail.core.shiro.PageAccessControl;
import uk.q3c.krail.core.shiro.ShiroVaadinModule;
import uk.q3c.krail.core.ui.DefaultUIModule;
import uk.q3c.krail.core.user.UserModule;
import uk.q3c.krail.core.view.PublicHomeView;
import uk.q3c.krail.core.view.ViewModule;
import uk.q3c.krail.core.view.component.DefaultComponentModule;
import uk.q3c.krail.i18n.DescriptionKey;
import uk.q3c.krail.i18n.LabelKey;
import uk.q3c.krail.testutil.TestI18NModule;
import uk.q3c.krail.testutil.TestOptionModule;
import uk.q3c.krail.testutil.TestPersistenceModule;
import uk.q3c.krail.util.DefaultResourceUtils;
import uk.q3c.krail.util.ResourceUtils;
import uk.q3c.krail.util.UtilsModule;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * This test uses all standard implementations to inject into {@link DefaultSitemapService}.
 *
 * @author David Sowerby
 */

@RunWith(MycilaJunitRunner.class)
@GuiceContext({TestDirectSitemapModule.class, UIScopeModule.class, ViewModule.class, EventBusModule.class, ServicesModule.class,
        ShiroVaadinModule.class, TestI18NModule.class, SitemapModule.class, UserModule.class, ApplicationConfigurationModule.class, DefaultShiroModule.class,
        DefaultComponentModule.class, TestPersistenceModule.class, StandardPagesModule.class, VaadinSessionScopeModule.class, TestOptionModule.class,
        NavigationModule.class, UtilsModule.class, DefaultUIModule.class})
public class DefaultSitemapServiceTest {

    static VaadinService vaadinService;
    static ResourceUtils resourceUtils = new DefaultResourceUtils();
    private final int FILE_NODE_COUNT = 4;
    private final int DIRECT_NODE_COUNT = 2;
    private final int STANDARD_NODE_COUNT = 5;
    @Inject
    DefaultSitemapService service;
    @Inject
    ServicesModel servicesModel;
    @Inject
    MasterSitemapQueue masterSitemapQueue;
    HierarchicalINIConfiguration iniConfig;

    @BeforeClass
    public static void setupClass() {
        vaadinService = mock(VaadinService.class);
        when(vaadinService.getBaseDirectory()).thenReturn(resourceUtils.userTempDirectory());
        VaadinService.setCurrent(vaadinService);
    }

    @Before
    public void setup() throws ConfigurationException {

        File inifile = new File(resourceUtils.userTempDirectory(), "WEB-INF/krail.ini");
        iniConfig = new HierarchicalINIConfiguration(inifile);
        iniConfig.clear();
        iniConfig.save();
        servicesModel.addService(service.getServiceKey());
    }

    @After
    public void teardown() throws Exception {
        service.stop();
        iniConfig.clear();
        iniConfig.save();
    }





    @Test
    public void nameAndDescription() {

        // given

        // when

        // then

        assertThat(service.getNameKey()).isEqualTo(LabelKey.Sitemap_Service);
        assertThat(service.getDescriptionKey()).isEqualTo(DescriptionKey.Sitemap_Service);
        assertThat(service.getName()).isEqualTo("Sitemap Service");
        assertThat(service.getDescription()).isEqualTo("This service creates the Sitemap using options from the " + "application configuration");
    }

    public void sourcesPropertyMissing() throws Exception {

        // given
        iniConfig.clear();
        iniConfig.save();

        // when
        service.start();
        // then
        assertThat(service.getReport()).isNotNull();
        assertThat(service.isStarted()).isTrue();
        assertThat(masterSitemapQueue.getCurrentModel()
                                     .getNodeCount()).isEqualTo(13);
        assertThat(service.getSourceTypes()).containsOnly();
    }

    /**
     * Key is there but value is not
     *
     * @throws Exception
     */
    public void sourcesPropertyEmpty() throws Exception {

        // given
        List<String> sources = new ArrayList<>();
        iniConfig.setDelimiterParsingDisabled(true);
        iniConfig.setProperty(ConfigKeys.SITEMAP_SOURCES, sources);
        iniConfig.setDelimiterParsingDisabled(false);
        iniConfig.save();

        // when
        service.start();
        // then
        assertThat(service.getReport()).isNotNull();
        assertThat(service.isStarted()).isTrue();
        assertThat(masterSitemapQueue.getCurrentModel()
                                     .getNodeCount()).isEqualTo(13);
        assertThat(service.getSourceTypes()).containsOnly();
    }

    @Test
    public void invalidSource_butValidOnePresent() throws Exception {
        // given
        List<String> sources = new ArrayList<>();
        sources.add("wiggly");
        sources.add("direct");
        iniConfig.setDelimiterParsingDisabled(true);
        iniConfig.setProperty(ConfigKeys.SITEMAP_SOURCES, sources);
        iniConfig.setDelimiterParsingDisabled(false);
        iniConfig.save();

        // when
        service.start();
        // then
        assertThat(service.getReport()).isNotNull();
        assertThat(service.isStarted()).isTrue();
        assertThat(masterSitemapQueue.getCurrentModel()
                                     .getNodeCount()).isEqualTo(7);
    }

    @Test
    public void stop() {

        // given

        // when

        // then
        assertThat(service.isLoaded()).isFalse();

    }

//

    public void invalidSource_noGoodOnes() throws Exception {

        // given
        List<String> sources = new ArrayList<>();
        sources.add("wiggly");
        sources.add("wobbly");
        iniConfig.setDelimiterParsingDisabled(true);
        iniConfig.setProperty(ConfigKeys.SITEMAP_SOURCES, sources);
        iniConfig.setDelimiterParsingDisabled(false);
        iniConfig.save();
        System.out.println(iniConfig.getProperty(ConfigKeys.SITEMAP_SOURCES));

        // when
        ServiceStatus status = service.start();
        // then
        assertThat(status.getState()).isEqualTo(Service.State.STOPPED);
        assertThat(status.getCause()).isEqualTo(Service.Cause.FAILED_TO_START);

    }



    public static class TestDirectSitemapModule extends DirectSitemapModule {

        @Override
        protected void define() {
            addEntry("direct", null, LabelKey.Home_Page, PageAccessControl.PUBLIC);
            addEntry("direct/a", PublicHomeView.class, LabelKey.Home_Page, PageAccessControl.PUBLIC);
            addRedirect("direct", "direct/a");
        }

    }



}

File: src/test/kotlin/uk/q3c/krail/core/navigate/sitemap/DefaultSitemapServiceTest.kt
package uk.q3c.krail.core.navigate.sitemap

import com.google.inject.Guice
import com.google.inject.Injector
import com.google.inject.Module
import io.mockk.every
import io.mockk.mockk
import org.amshove.kluent.shouldBe
import org.amshove.kluent.shouldContain
import org.amshove.kluent.shouldEqual
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.given
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on
import uk.q3c.krail.config.ApplicationConfiguration
import uk.q3c.krail.core.i18n.DescriptionKey
import uk.q3c.krail.core.i18n.Descriptions
import uk.q3c.krail.core.i18n.LabelKey
import uk.q3c.krail.core.navigate.StrictURIFragmentHandler
import uk.q3c.krail.core.navigate.URIFragmentHandler
import uk.q3c.krail.core.shiro.PageAccessControl
import uk.q3c.krail.core.vaadin.MockVaadinSession
import uk.q3c.krail.core.vaadin.createMockVaadinSession
import uk.q3c.krail.core.view.PublicHomeView
import uk.q3c.krail.eventbus.MessageBus
import uk.q3c.krail.i18n.CurrentLocale
import uk.q3c.krail.i18n.Translate
import uk.q3c.krail.i18n.test.MockCurrentLocale
import uk.q3c.krail.i18n.test.MockTranslate
import uk.q3c.krail.service.Cause
import uk.q3c.krail.service.State
import uk.q3c.krail.util.DefaultResourceUtils
import uk.q3c.krail.util.ResourceUtils
import uk.q3c.util.clazz.ClassNameUtils
import uk.q3c.util.clazz.DefaultClassNameUtils
import uk.q3c.util.guice.SerializationSupport
import uk.q3c.util.text.DefaultMessageFormat
import uk.q3c.util.text.MessageFormat2

/**
 * Created by David Sowerby on 10 May 2018
 */
object DefaultSitemapServiceTest : Spek({

    given("a Guice built sitemap service") {
        val resourceUtils: ResourceUtils = DefaultResourceUtils()
        lateinit var mockVaadinSession: MockVaadinSession
        lateinit var injector: Injector
        lateinit var config: ApplicationConfiguration
        lateinit var service: SitemapService
        lateinit var testModule: LocalTestModule

        beforeEachTest {
            testModule = LocalTestModule()
            mockVaadinSession = createMockVaadinSession()
            config = testModule.applicationConfiguration
            val modules = guiceModules()
            modules.add(testModule)
            injector = Guice.createInjector(modules)
            service = injector.getInstance(SitemapService::class.java)
        }
        afterEachTest {
            mockVaadinSession.clear()
        }

        on("service construction") {
            it("has name and description set") {
                service.nameKey.shouldEqual(LabelKey.Sitemap_Service)
                service.descriptionKey.shouldEqual(DescriptionKey.Sitemap_Service)
                service.name.shouldEqual("Sitemap Service")
                val descriptions = Descriptions()
                descriptions.keyClass = DescriptionKey::class.java
                descriptions.load()
                descriptions.map.get(service.descriptionKey).shouldEqual("This service creates the Sitemap using options from the application configuration")
            }
        }



        on("sources being empty") {
            every<List<String>> { config.getPropertyValue(SITEMAP_SOURCES, any()) } returns arrayListOf(SitemapSourceType.ANNOTATION.name, SitemapSourceType.DIRECT.name)
            service.start()

            it("uses default sources") {
                service.state.shouldEqual(State.RUNNING)
                service.sourceTypes.size.shouldBe(2)
                service.sourceTypes.shouldContain(SitemapSourceType.ANNOTATION)
                service.sourceTypes.shouldContain(SitemapSourceType.DIRECT)
            }
        }

        on("sources being defined") {
            every<List<String>> { config.getPropertyValue(SITEMAP_SOURCES, any()) } returns arrayListOf(SitemapSourceType.ANNOTATION.name)
            service.start()
            it("uses defined sources") {
                service.state.shouldEqual(State.RUNNING)
                service.sourceTypes.size.shouldBe(1)
                service.sourceTypes.shouldContain(SitemapSourceType.ANNOTATION)
            }
        }

        on("only invalid source defined") {
            every<List<String>> { config.getPropertyValue(SITEMAP_SOURCES, any()) } returns arrayListOf("rubbish")
            service.start()
            it("uses defined sources") {
                service.state.shouldEqual(State.FAILED)
                service.cause.shouldBe(Cause.FAILED_TO_START)
                service.sourceTypes.size.shouldBe(0)
            }
        }
    }

})

private fun guiceModules(): MutableList<Module> {
    return mutableListOf(TestDirectSitemapModule())
}

private class TestDirectSitemapModule : DirectSitemapModule() {

    override fun define() {
        addEntry("direct", LabelKey.Home_Page, PageAccessControl.PUBLIC)
        addEntry("direct/a", LabelKey.Home_Page, PageAccessControl.PUBLIC, PublicHomeView::class.java)
        addRedirect("direct", "direct/a")
    }
}

private class LocalTestModule : SitemapModule() {
    val applicationConfiguration: ApplicationConfiguration = mockk(relaxed = true)
    val messageBus: MessageBus = mockk(relaxed = true)
    val serialisationSupport: SerializationSupport = mockk()
    val userSitemap: UserSitemap = mockk(relaxed = true)

    override fun configure() {
        super.configure()
        bind(ApplicationConfiguration::class.java).toInstance(applicationConfiguration)
        bind(URIFragmentHandler::class.java).to(StrictURIFragmentHandler::class.java)
        bind(MessageBus::class.java).toInstance(messageBus)
        bind(CurrentLocale::class.java).toInstance(MockCurrentLocale())
        bind(Translate::class.java).toInstance(MockTranslate())
        bind(ClassNameUtils::class.java).to(DefaultClassNameUtils::class.java)
        bind(SerializationSupport::class.java).toInstance(serialisationSupport)
        bind(MessageFormat2::class.java).to(DefaultMessageFormat::class.java)
    }


    override fun bindUserSitemap() {
        bind(UserSitemap::class.java).toInstance(userSitemap)
    }
}


// UIScopeModule::class.java, ViewModule::class.java, VaadinEventBusModule::class.java,
//ShiroVaadinModule::class.java, TestKrailI18NModule::class.java, , UserModule::class.java, KrailApplicationConfigurationModule::class.java, DefaultShiroModule::class.java,
//DefaultComponentModule::class.java, InMemoryModule::class.java, StandardPagesModule::class.java, VaadinSessionScopeModule::class.java,
//NavigationModule::class.java, ServletEnvironmentModule::class.java, SerializationSupportModule::class.java, PushModule::class.java, EventBusModule::class.java, UtilsModule::class.java, UtilModule::class.java, DefaultUIModule::class.java, TestOptionModule::class.java